### ISSUE TITLE
Changes in our Git Branching flow

### ISSUE THREAD

**Issue by davigonz** [owncloud/android#2359]
Considering the following scenario with our current Git branching flow:

![current_branching_flow](https://user-images.githubusercontent.com/15261363/48621255-3a308700-e9a3-11e8-8674-a734e4a11b46.png)

1. We create branches to work on 2.9.2 features that are eventually merged into `master` branch.
2. Once 2.9.2 development is finished, we create `release_2.9.2` branch. At this point,`master` branch is blocked and if I or other contributors want to keep working on features for 2.10 release and merge them is not possible because we have to wait for merging and tagging `release_2.9.2`(step 3).
3. Once `release_2.9.2` is approved and fully tested, me merged it into master, tag the commit with `oc-android-2.9.2` and make the release public via GitHub and PlayStore. 
4. Merge `master` into `stable`.

This branching flow comes from long time ago and to speed up and not block our development process, including contributions, and according to what [GitFlow](https://es.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) defines, I would suggest the next changes: 

![new_branching_flow](https://user-images.githubusercontent.com/15261363/48621258-3b61b400-e9a3-11e8-83f3-90d48583892d.png)

1. Same as before
2. Once 2.9.2 development is finished, we create `release_2.9.2` branch but `master` is not blocked anymore and we can keep merging features for coming releases because of step 3 changes. 
3. Once `release_2.9.2` is approved and fully tested, me merged it into `stable`, tag the commit `oc-android-2.9.2` and make the release public via GitHub and PlayStore. 
4. Merge `stable` into `master`.

Think this workflow is widely used and fits properly to our case, what are your thoughts about? @felixboehm @PVince81 @michaelstingl @jesmrec

**Comment by michaelstingl** [comment]
@davigonz we use [OneFlow](https://www.endoflineblog.com/oneflow-a-git-branching-model-and-workflow) (simple version) in the [`ios-app`](https://github.com/owncloud/ios-app/). Could you also have a look?

**Comment by hannesa2** [comment]
When I read "At this point,master branch is blocked ..." this is really bad !
Maybe this is the reason why PR needs that much time here.

It's your decision what to do, but you should definite avoid blocking

**Comment by hannesa2** [comment]
I vote at least for a removement of the PR-must-based-on-recent-master rule and accept such a git flow
<img width="62" alt="image" src="https://user-images.githubusercontent.com/3314607/51959941-05d34f80-2457-11e9-8eef-0febf22a25ae.png">

Advantage: the annoying "please rebase on master, then we can merge" tasks are gone

**Comment by davigonz** [comment]
> I vote at least for a removement of the PR-must-based-on-recent-master rule
> Advantage: the annoying "please rebase on master, then we can merge" tasks are gone

Do you mean do not merge any feature till we release a new version? That's what I see in the git flow you attach.

**Comment by hannesa2** [comment]
Sorry bad example, I hope the clarify it
<img width="81" alt="image" src="https://user-images.githubusercontent.com/3314607/52002708-f4c02800-24c2-11e9-8e7b-32f65768d602.png">
PR-must-based-**NOT**-on-recent-master rule

Instead of current
<img width="33" alt="image" src="https://user-images.githubusercontent.com/3314607/52002795-29cc7a80-24c3-11e9-85ce-0c2ed3bd3def.png">
PR-must-based-on-recent-master rule

**Comment by davigonz** [comment]
So you propose to use merge instead of rebase, is that correct? Could you give us an example by using branch names like the ones we use here?

**Comment by hannesa2** [comment]
Correct, merge instead of rebase ! It avoids tons of "_please rebase on master_"

All development is done in master and after a `release_2.9.2` was created, only cherry picks to `release_2.9.2` will be done.

No merge `release_2.9.2` to `master` is valid. This cross merge makes confusion.

-my suggestion-

**Comment by hannesa2** [comment]
Ordinary I'm a rebaser, but when you work in a big team with a lot parallel epics, I mutated to a merger.
When here code format or lints are merged within hours, then I like rebase. But to be honest, this is not the reality (sorry, it's not to blame someone. It's just to explain)
What me make here to a merger, is that even simply commit needs weeks to month to be merged.

**Comment by michaelstingl** [comment]
Could the merge button options help here?

![options](https://user-images.githubusercontent.com/214010/52196928-f6d90c80-285d-11e9-878b-4422fa2ff5c1.png)

In the repo of the new iOS app, we only have the squash-merging button enabled.

**Comment by davigonz** [comment]
> When here code format or lints are merged within hours, then I like rebase.

IMHO, choosing rebase or merge is not a matter of having the features merged asap but a matter of what a team and a project needs. In our case, having an ordered history is what we wanted/needed till now and using rebase provides us that because we have the commits ordered by feature and not mixed in case we want to know where a specific issue comes from.

> Could the merge button options help here?

Maybe allow rebase merging? Documentation says: _This allows contributors to merge a pull request by rebasing their individual commits onto the base branch._

So I guess they do not need to perform the rebase manually any more but not sure what's going to happen when some conflicts come up.

**Comment by hannesa2** [comment]
Normally speed matters, but not everywhere. So you stay on recent behavior

**Comment by jesmrec** [comment]
Already moved to GitFlow.
